File: 0826-most-profit-assigning-work/0826-most-profit-assigning-work.py
class Solution:
    def maxProfitAssignment(self, difficulty: List[int], profit: List[int], worker: List[int]) -> int:
        job_profile = [(difficulty[i], profit[i]) for i in range(len(difficulty))]

        # Sort both worker and job_profile arrays
        job_profile.sort()
        worker.sort()

        total_profile, max_profile, index = 0, 0, 0

        for ability in worker:
            # While the index has not reached the end and worker can pick a job
            # with greater difficulty move ahead.
            while index < len(difficulty) and ability >= job_profile[index][0]:
                max_profile = max(max_profile, job_profile[index][1])
                index += 1
            total_profile += max_profile
        return total_profile

File: 0826-most-profit-assigning-work/0826-most-profit-assigning-work.cpp
class Solution {
public:
    int maxProfitAssignment(vector<int>& difficulty, vector<int>& profit, vector<int>& worker) {
        vector<pair<int, int>> jobProfile;
        for (int i = 0; i < difficulty.size(); i++){
            jobProfile.push_back({difficulty[i], profit[i]});
        }

        // Sort both worker and jobProfile arrays
        sort(worker.begin(), worker.end());
        sort(jobProfile.begin(), jobProfile.end());

        int totalProfile = 0, maxProfile = 0, index = 0;
        for (int i = 0; i < worker.size(); i++){
            // While the index has not reached the end and worker can pick a job
            // with greater difficulty move ahead.
            while (index < difficulty.size() && worker[i] >= jobProfile[index].first){
                maxProfile = max(maxProfile, jobProfile[index].second);
                index++;
            }
            totalProfile += maxProfile;
        }
        return totalProfile;
    }
};